### PR TITLE
Reuse CachedAppInfo interface to switch between legacy and config in code

### PR DESCRIPTION
**Pull request by rezaansyed**
### WHY are these changes introduced?

Resolves: https://github.com/Shopify/internal-cli-foundations/issues/761

### WHAT is this pull request doing?

Makes it so that `app dev` reads from the chosen configuration file. Also accepts a new `--config` option to choose another config.

### How to test your changes?

- `pnpm shopify app config link`, create a new app and call the config `production`
- `pnpm shopify app config link`, create a new app and call the config `staging`
- `pnpm shopify app dev` should pick up the staging config
- make a change to the config tile
- `pnpm shopify app config push`
- now make a change to `shopify.app.production.toml`
- `pnpm shopify app config push --config production`

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [x] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [x] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: packages/app/src/cli/models/app/app.ts [hunk 1]
```diff
@@ -19,11 +19,11 @@ const LegacyAppSchema = zod
...
   .object({
     name: zod.string(),
-    api_contact_email: zod.string(),
     client_id: zod.string(),
     scopes: zod.string().optional(),
-    webhook_api_version: zod.string(),
-    application_url: zod.string(),
+    api_contact_email: zod.string().optional(),
+    webhook_api_version: zod.string().optional(),
+    application_url: zod.string().optional(),
```
Comment (rezaansyed) [inline packages/app/src/cli/models/app/app.ts:26]: Hey @Arkham, what's the reason for making these optional?

Comment (Arkham) [reply packages/app/src/cli/models/app/app.ts:26]: we'd need to pass them everywhere, even in tests. what makes it super annoying is that if you forget to pass one of those the test will fail with no indication that it was the cause of failure.

Comment (rezaansyed) [reply packages/app/src/cli/models/app/app.ts:26]: But if they are required fields as part of app config upstream, should that not be reflected?

Comment (Arkham) [reply packages/app/src/cli/models/app/app.ts:26]: are they required? I thought all fields were optional except for `client_id`

Comment (Arkham) [reply packages/app/src/cli/models/app/app.ts:26]: let's address this in a follow-up PR, I just wanted to quickly unblock myself :)

**Review by Arkham — APPROVED** [review]
LGTM

**Pull request merged**